VS2017 and modernise C#

drewnoakes opened the pull request:
Following feedback on #4481, this is the first of several smaller PRs related to migration to VS2017 and the C# language features therein.

Changes proposed in this pull request:
 - Upgrade solution to VS2017
 - Set explicity `<LangVersion>` on all csproj files
 - Modernise C# code, taking advantage of more modern C# features

What did I do to test the code and ensure quality:
 - Applied changes systematically, using R# where possible for automation
 - Ran unit tests
 - Ran manual tests
 - Kept changes isolated between commits to aid review

Hopefully one or more reviewers will be willing to take on this subset of changes. Let me know of any changes you'd like to see.

It will be faster overall to review each commit separately. A given commit only contains one type of change, so it's much faster to browse through the diff. When they're combined it's more taxing to review, IMO.

Has been tested on:
 - GIT 2.16
 - Windows 10

gerhardol commented on GitCommands/GitCommands.csproj:
```diff
@@ -35,6 +35,7 @@
     <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
     <TargetFrameworkProfile>
     </TargetFrameworkProfile>
+    <LangVersion>7.2</LangVersion>
```
Since we are using the latest version, should we set the version here?

EbenZhang replied:
try `latest`

drewnoakes replied:
We can set this to `latest` to always be on the latest version.

But if @gerhardol's question was whether this attribute needs to be present at all, the answer is yes. By default, with a VS2017 project you compile using C# 7.0. The point releases are made optional, to reduce the obligation to test on newer compilers, which allows more regular compiler releases.

Differences between 7.2 and 7.0 are minor, but they do improve some aspects of the language.

drewnoakes replied:
I preferred being explicit about the version to reduce potential for problems where different developers are on different versions of the compiler.

RussKie replied:
I'm not against it, but my concern is maintenance overhead associated with the explicit specification.

As a thought, is there something we can set for the Appveyor to target (like 7.2) and keep "latest" in projects?

drewnoakes replied:
The benefit of being explicit in the source is that it applies to all environments equally, be it your PC, mine, or the CI server. If in future the project wants to support 7.3 features, a quick search/replace should update it. Developers wanting to add this to new projects would have to do so only if they need features newer than 7.0.

I believe you can pass the `-langversion` argument to `msbuild` if you want to make it explict on AppVeyor though.

sharwell replied:
❗️ *Never use latest*

Latest means everyone has a different version and there is no record of what the acceptable one actually is. The option exists for compatibility purposes only, and the only place it should ever be seen is in the compiler test suite that verifies it works.

gerhardol commented on GitCommands/Git/GitModule.cs:
```diff
@@ -60,7 +60,7 @@ public ConflictedFileData(string hash, string filename)
     }
 
-    [DebuggerDisplay("{Filename}")]
+    [DebuggerDisplay("{" + nameof(Filename) + "}")]
```
Is this the same? (Question, I am not familiar)

EbenZhang replied:
should be the same

drewnoakes replied:
Yes it's the same. The `DebuggerDisplayAttribute` evaluates a format string where member names are enclosed in braces, and shows the resulting string in the debugger for instances of the associated type.

By using `nameof` here, we are protected against breakages from automated refactoring tools that cannot introspect into the original format string.

RussKie replied:
Can this be `[DebuggerDisplay($"{nameof(Filename)}")]`

vbjay replied:
No. It can be 

``[DebuggerDisplay($"{{{nameof(Filename)}}}")]``

Interpolated strings escape {} with double {}

drewnoakes replied:
@vbjay you can't use `FormattableString` in attributes. Attribute arguments must be compile time constants. Even though the compiler could compute such a compile time value here, it doesn't. If you want to use `nameof`, the approach I have gone for is the only working route I know of.

vbjay replied:
Oops. Good point. I just saw the text wrong format of the interpolated
string.

vbjay replied:
Oops. Good point. I just saw the text wrong format of the interpolated
string.

RussKie replied:
No worries then, I'm fine with that.

sharwell replied:
➡️ Related to dotnet/csharplang#384

drewnoakes commented:
I encourage any reviewer to look at each commit in isolation. The `nameof` commit introduces the majority of the changes, for example. Others are much smaller, and more interesting from a review perspective anyway. The `nameof` commit was an automated refactoring, which did a fine job.

RussKie commented on GitCommands/Git/GitCommandCache.cs:
```diff
@@ -94,10 +94,8 @@ public static void Add(string cmd, byte[] output, byte[] error)
-                if (queue != null)
-                    queue.Clear();
-            }
+                queue?.Clear();
```
I don't want to lose `{ }`

In fact, I'd love to see them everywhere :sunglasses:

ericnewton76 replied:
@RussKie i kinda agree.  Sometimes terser doesnt make it easier to read it.

RussKie commented:
I've gone through the changes on GH and I'm fine with them, fully supportive of them.

My only concern is with explicit `LangVersion=7.2` - I don't to add any more maintenance than there is already. Any new project will need to be set explicitly and... it just won't happen.

drewnoakes commented:
> Any new project will need to be set explicitly 

Only if they want to use C# 7.1+ features.

### C# 7.1

- Async main
- `default` type inference
- inferred tuple element names
- pattern matching with generics

### C# 7.2

- `ref readonly`
- `private protected`
- digit separator after base specifier
- non-trailing named arguments

The odds of new projects using these are pretty low. They're nice in certain corner cases, but not critical. Anyone who expects to use these will know how to add the element to the project. In fact, ReSharper can do this for you and I wouldn't be surprised if there's a Roslyn analyser for it too, built in.

drewnoakes commented:
Just pushed up another commit (d9be2bb) that uses `default` type inference in a few places.

sharwell commented on GitCommands/AsyncLoader.cs:
```diff
@@ -79,8 +79,7 @@ public Task Load(Action loadContent, Action onLoaded)
             Cancel();
-            if (_cancelledTokenSource != null)
-                _cancelledTokenSource.Dispose();
+            _cancelledTokenSource?.Dispose();
```
:memo: Conversions of null check to `?.` should be sent as a single pull request with no other changes

sharwell commented on GitCommands/AsyncLoader.cs:
```diff
@@ -138,7 +136,7 @@ public Task<T> Load<T>(Func<CancellationToken, T> loadContent, Action<T> onLoade
                     if (token.IsCancellationRequested)
                     {
-                        return default(T);
+                        return default;
```
:memo: Conversions of `default(T)` to `default` should be sent as a single pull request with no other changes (may need to build off another pull request)

sharwell commented on GitCommands/CommitDataManager.cs:
```diff
@@ -66,7 +66,7 @@ public void UpdateCommitMessage(CommitData data, string sha1, ref string error)
             var module = GetModule();
             if (sha1 == null)
-                throw new ArgumentNullException("sha1");
+                throw new ArgumentNullException(nameof(sha1));
```
:memo: Conversion of string literals to `nameof` should be sent as a single pull request with no other changes

sharwell commented on GitCommands/CommitTemplateItem.cs:
```diff
@@ -56,8 +56,7 @@ public static void SaveToSettings(CommitTemplateItem[] items)
             string serializedString = AppSettings.CommitTemplates;
-            bool shouldBeUpdated;
-            var templates = DeserializeCommitTemplates(serializedString, out shouldBeUpdated);
+            var templates = DeserializeCommitTemplates(serializedString, out var shouldBeUpdated);
```
:memo: Refactoring to use `out T` should be sent as a single pull request with no other changes

sharwell commented on GitCommands/Config/ConfigFile.cs:
```diff
@@ -10,8 +10,7 @@ namespace GitCommands.Config
     {
-        private readonly string _fileName;
-        public string FileName { get { return _fileName; } }
+        public string FileName { get; }
```
:memo: Using getter-only auto-properties should be sent as a separate pull request

sharwell commented on GitCommands/Config/ConfigFile.cs:
```diff
@@ -296,8 +292,9 @@ private class ConfigFileParser
             private IConfigSection _section = null;
-            private string FileName { get { return _configFile._fileName; } }
             private string _key = null;
+            private string FileName => _configFile.FileName;
```
:memo: Conversion to expression-bodied members should be sent as a separate pull request

sharwell commented on GitCommands/GitCommands.csproj:
```diff
@@ -35,6 +35,7 @@
     <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
     <TargetFrameworkProfile>
     </TargetFrameworkProfile>
+    <LangVersion>7.2</LangVersion>
```
:memo: This should be set in a **Directory.Build.props** file that gets used by all projects automatically.

sharwell commented:
> Upgrade solution to VS2017

On this specifically, if y'all are interested in a move in this direction I can convert the solution to the new project system so the project files don't have to be modified in the future when files are added/moved.

drewnoakes commented:
@sharwell thanks for your review.

> separate pull requests

These commits are already a subdivision of #4481. The features you'd split out are individual commits. So, if there's a part of this PR that you wouldn't want to merge, I can just remove that one commit. Given the dependency between commits, it's much easier this way. If this PR is merged, I will look at bringing more modernisation things over from #4481 too as separate PRs. Do you really want 50+ PRs? I personally don't have the steam to do that, as they'd each have to happen one after the other to prevent merge conflicts. If you still feel splitting them up is important, would you explain why?

> `Directory.Build.props`

This sounds interesting. I'll take a look, thanks.

- I can convert the solution to the new project system

I tried this once for a WPF project and hit a hurdle, though I can't recall now what it was though I remember thinking it applied to WinForms too. It might have been [this](https://github.com/dotnet/project-system/issues/159). Do you know if newer SDK projects support all project types? Could the the WiX installer project be converted, for example? Having some projects using one format and others another might be more confusing than useful. What other benefits do you see?

drewnoakes commented:
@sharwell @RussKie I've edited the PR to use a `Directory.Build.props` file which means the `<LangVersion>` attribute applies to all projects and need only be specified in one place. Much nicer. Thanks for the suggestion.

vbjay commented:
Oops. Good point. I just saw the text wrong format of the interpolated
string.

On Thu, Feb 22, 2018, 7:49 AM Drew Noakes <notifications@github.com> wrote:

> *@drewnoakes* commented on this pull request.
> ------------------------------
>
> In GitCommands/Git/GitModule.cs
> <https://github.com/gitextensions/gitextensions/pull/4507#discussion_r169944909>
> :
>
> > @@ -60,7 +60,7 @@ public ConflictedFileData(string hash, string filename)
>          public string Filename;
>      }
>
> -    [DebuggerDisplay("{Filename}")]
> +    [DebuggerDisplay("{" + nameof(Filename) + "}")]
>
> @vbjay <https://github.com/vbjay> you can't use FormattableString in
> attributes. Attribute arguments must be compile time constants. Even though
> the compiler could compute such a compile time value here, it doesn't. If
> you want to use nameof, the approach I have gone for is the only working
> route I know of.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4507#discussion_r169944909>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsQqa4NeI46lXBYIUF1xUFCQ5JlIpks5tXWJngaJpZM4SOcD4>
> .
>

sharwell commented on ResourceManager/GitExtensionsUserControl.cs:
```diff
@@ -132,16 +129,12 @@ protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
         {
-            var hotkey = GetHotkeyCommand(commandCode);
-            return hotkey == null ? Keys.None : hotkey.KeyData;
+            return GetHotkeyCommand(commandCode)?.KeyData ?? Keys.None;
```
:memo: Same here about `Keys` being a class/struct

drewnoakes replied:
`enum`

sharwell commented on GitUI/UserControls/RevisionGridClasses/DvcsGraph.Lanes.cs:
```diff
@@ -490,8 +489,7 @@ public void Expand(int col)
                         {
-                            int start, end;
-                            Graph.LaneInfo info = edges.RemoveNext(i, 0, out start, out end);
+                            Graph.LaneInfo info = edges.RemoveNext(i, 0, out var start, out var _);
```
:bulb: `out _`

drewnoakes replied:
Updated, thanks. Less is more.

sharwell commented on GitUI/UserControls/RevisionGridClasses/DvcsGraph.Lanes.cs:
```diff
@@ -502,8 +500,7 @@ public void Replace(int aOld, int aNew)
                     {
-                        int start, end;
-                        Graph.LaneInfo info = edges.RemoveNext(aOld, j, out start, out end);
+                        Graph.LaneInfo info = edges.RemoveNext(aOld, j, out var start, out var _);
```
:bulb: Can simplify to `out _`

drewnoakes replied:
Ditto

sharwell commented on GitUI/UserControls/RevisionGridClasses/DvcsGraph.cs:
```diff
@@ -1236,39 +1236,22 @@ public bool IsRevisionRelative(string aGuid)
-            }
-
-            return null;
+            return guid != null && _graphData.Nodes.TryGetValue(guid, out var node) ? (int?)node.Index : null;
```
💭 To me this is shorter, but I'm not sure it's simpler

sharwell commented:
* ⚠️ 0b91e17 (replace lambda with method group) represents a likely performance hit due to dotnet/roslyn#5835. I would not consider it a clear win so it should be evaluated separately.
* 29a7c07 (use `nameof`) should be fine
* ⚠️ I'm not a fan of b86a403 (join null check with assignment). It has weird impacts on code coverage metrics and can be a problem for debugging per dotnet/roslyn#22015.
* 4ce53b3 (use `??` expressions) seems fine
* a7042e2 (use auto properties) seems fine; note that formatting was botched in AffixEntry.cs, AffixRule.cs, PhoneticRule.cs, ReplaceWordEventArgs.cs, SpellingEventArgs.cs, but maybe that's fine
* 408c53b (use null propagation) seems fine
* d78faf1 (use base class qualifier) seems fine
* c3dff7d (remove unused usings) seems fine
* 1a8aaa2 (use `default` instead of `default(T)`) seems fine
* I'm on the fence about 1ce7013, mostly because the change uses `var` instead of explicit type names in a bunch of places where it didn't previously. If it matches the desired type style for the project then this is fine.

sharwell commented on GitExtensions.sln:
```diff
@@ -1,6 +1,6 @@
-VisualStudioVersion = 14.0.25123.0
+# Visual Studio 15
+VisualStudioVersion = 15.0.27130.2026
 MinimumVisualStudioVersion = 10.0.40219.1
```
💡 You can set the minimum version here to 15.5 and it will prevent the solution from opening in a version of VS lower than the first one that supports C# 7.2. The improved error message experience can be seen in https://github.com/Microsoft/perfview/pull/319#issuecomment-318511183.

drewnoakes replied:
Do you know if the revision/build values are important? Can we just use `15.5` or `15.5.5`?

sharwell replied:
It should work fine to use `15.5` 👍

drewnoakes replied:
`MinimumVisualStudioVersion = 15.5` gives me an error, as does 15.5.5.

![image](https://user-images.githubusercontent.com/350947/36548785-1618ea38-17e9-11e8-8451-4735f31f0aa6.png)

I have 15.5.5 installed. Not sure how to apply your suggestion here.

sharwell replied:
Oh, duh, one second...

sharwell replied:
15.5 is internally recognized as **15.0.27130.0**, so try that 😁

drewnoakes commented on GitCommands/Repository/RecentRepoInfo.cs:
```diff
@@ -34,10 +34,9 @@ public RecentRepoInfo(Repository aRepo, bool aMostRecent)
+            DirInfo = DirInfo?.Parent;
 
-            DirName = DirInfo == null ? "" : DirInfo.FullName;
+            DirName = DirInfo?.FullName ?? "";
```
`FullName` cannot be `null` so there is no semantics change here.

drewnoakes commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -2126,21 +2125,18 @@ private void GetUserSettings()
+            ListView lv = menu?.SourceControl as ListView;
 
-            ListView lv = menu.SourceControl as ListView;
             if (lv == null)
```
Thanks. Will update.

drewnoakes commented on GitUI/CommandsDialogs/SettingsDialog/Pages/ControlHotkeys.cs:
```diff
@@ -106,7 +106,7 @@ private void UpdateListViewItems(HotkeySettings setting)
         private void UpdateTextBox(HotkeyCommand command)
         {
-            txtHotkey.KeyData = (command != null) ? command.KeyData : Keys.None;
+            txtHotkey.KeyData = command?.KeyData ?? Keys.None;
```
`Keys` is an `enum`, and `KeyData` is not nullable. So no change here.

drewnoakes commented on GitUI/Script/SplitButton.cs:
```diff
@@ -471,7 +470,7 @@ private Size CalculateButtonAutoSize()
             Size ret_size = Size.Empty;
             Size text_size = TextRenderer.MeasureText(Text, Font);
-            Size image_size = Image == null ? Size.Empty : Image.Size;
+            Size image_size = Image?.Size ?? Size.Empty;
```
The `Size` property returns a `Size` struct, so no change here.

drewnoakes commented on Plugins/Bitbucket/GetPullRequest.cs:
```diff
@@ -17,7 +17,7 @@ public static PullRequest Parse(JObject json)
                 State = json["state"].ToString(),
                 Title = json["title"].ToString(),
-                Description = json["description"] != null ? json["description"].ToString() : "",
+                Description = json["description"]?.ToString() ?? "",
```
Yes, `ToString` should not return `null`. In the very unlikely event that it did, `Description` would be an empty string instead of null.

drewnoakes commented on ResourceManager/GitExtensionsFormBase.cs:
```diff
@@ -45,16 +45,12 @@ protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
         {
-            var hotkey = GetHotkeyCommand(commandCode);
-            return hotkey == null ? Keys.None : hotkey.KeyData;
+            return GetHotkeyCommand(commandCode)?.KeyData ?? Keys.None;
```
`enum`

drewnoakes commented:
>  0b91e17 (replace lambda with method group) represents a likely performance hit due to dotnet/roslyn#5835. I would not consider it a clear win so it should be evaluated separately.

Interesting. I didn't know about the lack of caching. Can drop this commit out if it's easier. Not a biggie either way I guess.

>  I'm not a fan of b86a403 (join null check with assignment). It has weird impacts on code coverage metrics and can be a problem for debugging per dotnet/roslyn#22015.

Can also drop this commit if it's easier. Does the project have any coverage metrics at the moment? Isn't that more of a bug in the coverage tools? Also with the debugger. If it's a big enough problem to avoid then that's fine with me.

My personal heuristic with these things is to reduce the amount of code so long as it doesn't mangle readability. Most of the changes here are in that light.

sharwell commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -2727,25 +2722,16 @@ protected override void Dispose(bool disposing)
-                    components.Dispose();
-                if (_gitStatusMonitor != null)
-                    _gitStatusMonitor.Dispose();
+                _commitButton?.Dispose();
```
Replacement analyzers will be available soon, e.g. dotnet/roslyn-analyzers#1595

sharwell commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -2150,9 +2140,7 @@ private void ViewFileHistoryMenuItem_Click(object sender, EventArgs e)
                 return;
 
             if (list.SelectedItems.Count() == 1)
-            {
```
It would be easy to implement this as a follow-up change. I have a set of analyzers which can enforce several different policies related to the preferred code style for braces, and then fix the entire code base to match those settings.

RussKie commented:
StyleCop? :D

On 24/02/2018 6:54 PM, "Sam Harwell" <notifications@github.com> wrote:

> *@sharwell* commented on this pull request.
> ------------------------------
>
> In GitUI/CommandsDialogs/FormCommit.cs
> <https://github.com/gitextensions/gitextensions/pull/4507#discussion_r170414935>
> :
>
> > @@ -2150,9 +2140,7 @@ private void ViewFileHistoryMenuItem_Click(object sender, EventArgs e)
>                  return;
>
>              if (list.SelectedItems.Count() == 1)
> -            {
>
> It would be easy to implement this as a follow-up change. I have a set of
> analyzers which can enforce several different policies related to the
> preferred code style for braces, and then fix the entire code base to match
> those settings.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4507#discussion_r170414935>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXoWMg0cmqiQCtUdA5S9xioUw9myrks5tX8A1gaJpZM4SOcD4>
> .
>

sharwell commented on Directory.Build.props:
```diff
@@ -0,0 +1,5 @@
+<Project>
+  <PropertyGroup>
+    <LangVersion>7.2</LangVersion>
```
In my experience, the best C# 7.2 feature for normal work is non-trailing named arguments. This feature allows literals (e.g. `true`, `null`, or `2`) passed as an argument in the middle of the list to get a name, without requiring everything that follows to have names as well.

If we ever have highly performance-sensitive code to deal with, some of the ref type changes could help. Nothing stood out as obviously required to me.

sharwell commented:
> StyleCop

StyleCop Analyzers is not the StyleCop people know. It's *extremely* fast, *extremely* well-tested, and instead of just complaining it fixes things for you.

RussKie commented:
I like the sound of it and am mighty interested

On 24/02/2018 6:58 PM, "Sam Harwell" <notifications@github.com> wrote:

> StyleCop
>
> StyleCop Analyzers is not the StyleCop people know. It's *extremely*
> fast, *extremely* well-tested, and instead of just complaining it fixes
> things for you.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4507#issuecomment-368209378>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXsQ7unDqLLyhbyNpMzjG3CowimK2ks5tX8EwgaJpZM4SOcD4>
> .
>

drewnoakes commented:
As requested:

- Removed 0b91e17 (Replace lambda with method group)
- Remove b86a403 (Join null check with assignment)
- Reverted null propagation in `Dispose` methods that created (invalid) diagnostic warnings (99a132c)
- Increased minimum VS version to 15.5 (4aa5b8d)

Is there anything else? What was the decision on C# 7.2 vs 7.0? Supporting 7.0 conflicts with the advice to increase the minimum VS version.

On the topic of style issues like braces, member ordering and whatnot, it would less confusing to contributors if this were systematised. The current codebase does not serve to educate on what's expected, nor is there any documentation. Documentation tends to be hard to follow and enforce -- tooling is a much better option. StyleCop Analysers sounds good, though I haven't tried it. It's possible to configure ReSharper via the `.sln.DotSettings` file to capture many if not all such policies.

But then of course there needs to be agreement on the policies themselves. Probably one for the BDFL's. I'd suggest considering rules that result in the fewest changes to existing code.

RussKie commented:
I haven't looked at the diff in details but reading the conversation I think I am happy with the proposed.

@gerhardol: I believe your concern has been addressed. Are you happy to proceed?

@jbialobr @KindDragon any thoughts?

jbialobr commented on Plugins/BuildServerIntegration/TeamCityIntegration/TeamCityAdapter.cs:
```diff
@@ -75,7 +75,7 @@ private CookieContainer GetTeamCityNtlmAuthCookie (string serverUrl, IBuildServe
             string url = serverUrl + "ntlmLogin.html";
             var cookieContainer = new CookieContainer();
-            var request = (HttpWebRequest)HttpWebRequest.Create (url);
+            var request = (HttpWebRequest)WebRequest.Create (url);
```
I don't understand this change. How WebRequest can be cast to HttpWebRequest?

drewnoakes replied:
There is no actual change to the generated IL here. Class `HttpWebRequest` derives from `WebReqeust`, but the method is static, so not overridden. This change results in the method being referenced via the type that actually declares it.

The method's return type depends upon the URI it's given. If you give it an `http://` URI, it'll return `HttpWebRequest`.

jbialobr replied:
Thanks, I code a lot in Pascal, so I unconsciously interpreted Create as a constructor :)

jbialobr commented:
> It would be easy to implement this as a follow-up change. I have a set of
 analyzers which can enforce several different policies related to the
 preferred code style for braces, and then fix the entire code base to match
 those settings.

I like the idea that code should be formatted automatically on save action. VS Code does it for our typescript projects very well.

drewnoakes commented:
> I like the idea that code should be formatted automatically on save action. VS Code does it for our typescript projects very well.

I created #4534 for tracking auto-formatting.

RussKie commented:
Hi Drew, can we please do a bit of house-keeping:
* Merge 9f05a43acda5e81f23728ca0a4f67cb94a13e4b9, dfd523b0fbfe58a449d6ce35ea7b8648d6cc4b35, b3117ace30796b53e0a5bd09a6200266caff47ec and  4aa5b8d604ca6770334be89361c40a379f9daadd together (the first three and the last)
* Remove 99a132c29f26e4b08c82c692a297672c102118f4 completely

And take the rest as they are.

Thank you

RussKie commented:
@sharwell are you happy with the changes?

drewnoakes commented:
@RussKie done.

RussKie commented:
Thank you
